src/services/exportInfoMap.ts

```diff
@@ -445,15 +447,32 @@ export function forEachExternalModuleToImportFrom(
     }
 }
 
-function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
-    const isExcluded = excludePatterns && ((fileName: string) => excludePatterns.some(p => p.test(fileName)));
+function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, host: LanguageServiceHost, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
+    const realpathsWithSymlinks = host.getSymlinkCache?.().getSymlinkedDirectoriesByRealpath();
+    const isExcluded = excludePatterns && (({ fileName, path }: SourceFile) => {
+        if (excludePatterns.some(p => p.test(fileName))) return true;
+        if (realpathsWithSymlinks?.size && pathContainsNodeModules(fileName)) {

```

Comment:
This `pathContainsNodeModules` check may be too conservative. The real world cases I’ve seen all have both realpaths and symlink paths somewhere inside a directory called `node_modules`. But if there are any package manager strategies where the realpaths live in a directory that _doesn’t_ contain `/node_modules/`, this check will fail.

Reply:
Checked yarn pnp and it contains both `node_modules` and the real/original package name, so that one’s fine. I think this logic is ok.

<img width="645" alt="image" src="https://github.com/microsoft/TypeScript/assets/3277153/fbc33b46-929a-4617-95e1-769241b73cd9">

Reply:
Doesnt npm link let your realpath be "say c:/typescript" from say "c:/myproject/node_modules/typescript" ? Or is that not what you are checking here?

Reply:
Hm, yes, I think that’s a case where this would fail. I’m not sure how important that is though. If you’ve bothered to npm link something you _probably_ want to be able to auto-import from it? I was hoping to have some sort of filter here for performance reasons.